kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaAuthnHandler.java

```diff
@@ -0,0 +1,403 @@
+                        onSaslAuthenticateRequest(ctx, (DecodedRequestFrame<SaslAuthenticateRequestData>) frame);
+                        return;
+                    default:
+                        if (lastSeen == State.AUTHN_SUCCESS) {
+                            ctx.fireChannelRead(msg);
+                        }
+                        else {
+                            ctx.writeAndFlush(
+                                    new DecodedResponseFrame<>(
+                                            frame.apiVersion(),
+                                            frame.correlationId(),
+                                            new ResponseHeaderData().setCorrelationId(frame.correlationId()),
+                                            errorResponse(frame, new IllegalSaslStateException("Nope"))
+                                    // TODO add support for session lifetime/reauth
+                                    ));
+                        }
+                }
+            }
+        }
+        else {
+            throw new IllegalStateException("Unexpected message " + msg.getClass());
+        }
+    }
+
+    private ApiMessage errorResponse(DecodedRequestFrame<?> frame, Throwable error) {
+        ApiMessage body;
+        switch (frame.apiKey()) {
+            case METADATA:
+                body = new MetadataRequest((MetadataRequestData) frame.body(), frame.apiVersion()).getErrorResponse(0, error).data();
+                break;
+            // TODO all the other cases!
+            default:
+                throw new IllegalStateException();
+        }
+        return body;
+    }
+
+    private void onSaslHandshakeRequest(ChannelHandlerContext ctx,
+                                        DecodedRequestFrame<SaslHandshakeRequestData> data)
+            throws SaslException {
+        String mechanism = data.body().mechanism();
+        Errors error;
+        if (lastSeen == State.AUTHN_SUCCESS) {
+            error = Errors.ILLEGAL_SASL_STATE;
+        }
+        else if (enabledMechanisms.contains(mechanism)) {
+            var cbh = mechanismHandlers.get(mechanism);
+            // TODO use SNI to supply the correct hostname
+            saslServer = Sasl.createSaslServer(mechanism, "kafka", null, null, cbh);
+            if (saslServer == null) {
+                throw new IllegalStateException("SASL mechanism had no providers: " + mechanism);
+            }
+            else {
+                error = Errors.NONE;
+            }
+        }
+        else {
+            error = Errors.UNSUPPORTED_SASL_MECHANISM;
+        }
+
+        ctx.writeAndFlush(new DecodedResponseFrame<>(
+                data.apiVersion(),
+                data.correlationId(),
+                new ResponseHeaderData().setCorrelationId(data.correlationId()),
+                new SaslHandshakeResponseData()
+                        .setMechanisms(enabledMechanisms)
+                        .setErrorCode(error.code())));
```

Comment:
should we pull out a method to do this? Something like 

```java
write(ctx, data, new SaslHandshakeResponseData()
                        .setMechanisms(enabledMechanisms)
                        .setErrorCode(error.code()))
``` 

```java
public static <R, B> write(ChannelHandlerContext ctx, DecodedRequestFrame<R> requestFrame, B Body) {
        ctx.writeAndFlush(new DecodedResponseFrame<>(
                data.apiVersion(),
                data.correlationId(),
                new ResponseHeaderData().setCorrelationId(data.correlationId()),
                body);
}

Reply:
Good spot.